pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podstore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	DeleteQueueName = "pod_delete"
+	PodIPIndex      = "PodIP"
+	DelayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	curPod       cache.Indexer
+	prevPod      cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+	// Mapping pod.uuid to LocalClock
+	timestampMap map[string]*LocalClock
+	clock        clock.WithTicker
+	mutex        sync.Mutex
+}
+
+type LocalClock struct {
+	// The local clock when the Pod is added
+	CreationTimestamp time.Time
+	// The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
+	DeletionTimestamp *time.Time
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPod:       podInformer.GetIndexer(),
+		prevPod:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+		clock:        clock,
+		timestampMap: map[string]*LocalClock{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			DeleteFunc: s.onPodDelete,
+		})
+	s.curPod.AddIndexers(cache.Indexers{PodIPIndex: podIPIndexFunc})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Error while processing Create Event")
```

Comment:
is this a bit too generic for an error message?

Reply:
Fixed. Thanks.